### PR TITLE
Fix inconsistent 24h volume on liquidity page

### LINKED ISSUE #1040: Inconsistent daily volume on liquidity page and dxstats
**Description**
I've just noticed that the daily volume on Ethereum, Arbitrum and Gnosis is not the same as on the dxstats for most of the pairs.

#### Definition of done
- [X] Description is created (Task creator)
- [x] Code is complete (Developer)
- [ ] Scope of test and what was done section is defined in task (Developer)
- [ ] Review passed (Code reviewer)
- [ ] Verified manually (QA)
- [ ] Test cases created/updated if necessary (QA)
- [ ] Change successfully merged
- [ ] Release notes added if needed

**Screenshots**

### PR DESCRIPTION
**Pull request by Diogomartf**
Fixes #1040 

Fix inconsistent 24h volume on liquidity page to match Swapr Stats.
- Add blocks subgraphs clients 
- Add new subgraph queries
- update 24h volume hook

image

# Problem

In **Swapr** liquidity pools we were fetching the daily volume from a subgraph but the daily volume is only of that day and not the last 24 hours so it was not accurate. That's why **DxStats** showed more volume.

# Summary

As I went through this I noticed we cannot filter data by date (timestamp) on Swapr subgraph. Time-travel queries are great for comparing statistics across time frames, the solution is to turn a timestamp into a block number using for eg. Ethereum Blocks subgraph ([read more](https://blocklytics.org/blog/ethereum-blocks-subgraph-made-for-time-travel/)) and then filter volume by that block. Like **DxStats** is doing.

I will use the pair **ETH/DAI** address as an example: `0x7515be43d16f871588adc135d58a9c30a71eb34f`

### Fetch block by timestamp

First, we fetch a block by timestamp in our case is 24 hours ago (_for eg._ `2022-08-03 22:59:15`  which is `1659563955`)

image

we get a `block` object with a `number`, eg: `15272017`

image

[check query getBlockByTimestamp on thegraph (blocklytics/ethereum-blocks)](https://api.thegraph.com/subgraphs/name/blocklytics/ethereum-blocks/graphql?query=query+getBlockByTimestamp+%7B%0A++blocks%28%0A++++first%3A+1%0A++++orderBy%3A+number%0A++++orderDirection%3A+asc%0A++++where%3A+%7Btimestamp_gt%3A+15272017%7D%0A++%29+%7B%0A++++number%0A++++__typename%0A++%7D%0A%7D)

### Fetch volume till that block

Then, we can fetch volume and filter with that `block`

image

we get a `pair` object with `volumeUSD`

image

[check query volumePairTillBlock on thegraph (dxgraphs/swapr-mainnet-v2)](https://api.thegraph.com/subgraphs/name/dxgraphs/swapr-mainnet-v2/graphql?query=query+volumePairTillBlock%28%24pair%3A+String%21%2C+%24block%3A+Int%21%29+%7B%0A++++pairs%28first%3A+1%2C+where%3A+%7B+id%3A+%24pair+%7D%2C+block%3A+%7B+number%3A+%24block+%7D%29+%7B%0A++++++volumeUSD%0A++++%7D%0A++%7D)

### Fetch the total volume

Now that we have the volume till a block from 24h ago we need the total volume to make the difference.

image

And we get again a `pair` object with the `volumeUSD` of that pair, as we are not filtering by block we get the total volume.

image

[check pairTotalVolume query on thegraph (dxgraphs/swapr-mainnet-v2)](https://api.thegraph.com/subgraphs/name/dxgraphs/swapr-mainnet-v2/graphql?query=query+pairTotalVolume%28%24pair%3A+String%21%29+%7B%0A++pairs%28first%3A+1%2C+where%3A+%7Bid%3A+%24pair%7D%29+%7B%0A++++volumeUSD%0A++%7D%0A%7D)

### 24 hours volume

`volume24h = totalVolume - volumeTill24hAgoBlock`

_eg, currently on ETH/DAI pair: 26442072 - 26401359 =  40713_

# To Test

1. Go to liquidity page
	- Expect prices to match [DxStats](https://dxstats.eth.limo/#/pairs)

### CODE REVIEW

**Comment by MilanVojnovic95** [comment]
Tested, and it is good everywhere except for Arbitrum, on swapr it shows 0 everywhere and on dxstats differently.

**Comment by Diogomartf** [comment]
> Tested, and it is good everywhere except for Arbitrum, on swapr it shows 0 everywhere and on dxstats differently.

Thanks, will take a look on that.

**Comment by Diogomartf** [comment]
> Tested, and it is good everywhere except for Arbitrum, on swapr it shows 0 everywhere and on dxstats differently.

It should be fixed now 💪

**Comment by niemam29** [comment]
Tested and looks good to me!

File: src/hooks/usePairVolume24hUSD.ts [hunk 1]
```diff
@@ -6,52 +6,142 @@ import Decimal from 'decimal.js-light'
...
-    if (loading) return { loading: true, volume24hUSD: ZERO_USD }
+    if (volumeTill24agoLoading || totalVolumeLoading || blockLoading) return { loading: true, volume24hUSD: ZERO_USD }
     if (
-      !data ||
-      (!isToken && (!data.pairDayDatas || data.pairDayDatas.length === 0 || !data.pairDayDatas[0])) ||
-      (isToken && (!data.tokenDayDatas || data.tokenDayDatas.length === 0 || !data.tokenDayDatas[0])) ||
-      error
+      !volumeTill24agoData ||
+      !totalVolumeData ||
+      volumeTill24agoData.pairs.length === 0 ||
```
Comment (mix1o) [inline src/hooks/usePairVolume24hUSD.ts:123]: small thing 
```suggestion
      !volumeTill24agoData.pairs.length ||
```

File: src/hooks/usePairVolume24hUSD.ts [hunk 2]
```diff
@@ -6,52 +6,142 @@ import Decimal from 'decimal.js-light'
...
+      totalVolumeData.pairs.length === 0 ||
```
Comment (mix1o) [inline src/hooks/usePairVolume24hUSD.ts:124]:
```suggestion
      !totalVolumeData.pairs.length ||
```

File: src/hooks/usePairVolume24hUSD.ts (hunk 1 above)
Comment (Diogomartf) [reply src/hooks/usePairVolume24hUSD.ts:123]: cleaner, thanks!

File: src/hooks/usePairVolume24hUSD.ts [hunk 3]
```diff
@@ -6,52 +6,142 @@ import Decimal from 'decimal.js-light'
...
   pairOrTokenAddress?: string | null,
   isToken = false
 ): { loading: boolean; volume24hUSD: CurrencyAmount } {
-  const { loading, data, error } = useQuery(isToken ? TOKENQUERY : QUERY, {
+  const { block, loading: blockLoading, error: blockError } = useBlockByTimestamp(getTimestamp24hAgo())
+
+  const {
+    data: volumeTill24agoData,
+    loading: volumeTill24agoLoading,
+    error: volumeTill24agoError,
```
[the remark this answers is not in the document]
Comment (Diogomartf) [reply src/hooks/usePairVolume24hUSD.ts:95]: Makes sense, will update it.

**Review by mix1o — APPROVED** [review]

**Comment by MilanVojnovic95** [comment]
Retested and looks ok to me 👍

**Comment by niemam29** [comment]
Retested and it's good!

File: src/apollo/client.ts [hunk 4]
```diff
@@ -74,3 +69,42 @@ export const carrotSubgraphClient: { [chainId: number]: ApolloClient<NormalizedC
...
+  [ChainId.ARBITRUM_RINKEBY]: 'https://api.thegraph.com/subgraphs/name/dodoex/arbitrum-one-blocks',
+  [ChainId.ARBITRUM_GOERLI]: '', // FIXME: fix this once the subgraph is deployed
+}
+
+export const subgraphBlocksClients: {
+  [chainId in SWPRSupportedChains]: ApolloClient<NormalizedCacheObject>
+} = {
+  [ChainId.MAINNET]: new ApolloClient({
+    uri: subgraphBlocksClientsUris[ChainId.MAINNET],
+    cache: new InMemoryCache(),
```
Comment (jorgedclopes) [inline src/apollo/client.ts:88]: optional: don't you want to make this pattern into a function and apply it to a list of ChainIds?

File: src/hooks/usePairVolume24hUSD.ts [hunk 5]
```diff
@@ -6,52 +6,142 @@ import Decimal from 'decimal.js-light'
...
+import { subgraphBlocksClients } from '../apollo/client'
 import { ZERO_USD } from '../constants'
+import { useActiveWeb3React } from '../hooks'
+import { SWPRSupportedChains } from '../utils/chainSupportsSWPR'
 
-const QUERY = gql`
-  query getPair24hVolume($pairOrTokenAddress: ID!, $date: Int!) {
-    pairDayDatas(first: 1, where: { pairAddress: $pairOrTokenAddress, date: $date }) {
-      dailyVolumeUSD
+const GET_BLOCK_BY_TIMESTAMP = gql`
```
Comment (jorgedclopes) [inline src/hooks/usePairVolume24hUSD.ts:14]: how many more queries are we going to have? At what point should we make a dict out of these?

File: src/apollo/client.ts (hunk 4 above)
Comment (Diogomartf) [reply src/apollo/client.ts:88]: hmm.. I'll take a look 👍

File: src/hooks/usePairVolume24hUSD.ts (hunk 5 above)
Comment (jorgedclopes) [reply src/hooks/usePairVolume24hUSD.ts:14]: a map, a {key, value} object.

File: src/apollo/client.ts (hunk 4 above)
Comment (Diogomartf) [reply src/apollo/client.ts]: Updated it. Thanks for the tip, much better now.

File: src/hooks/usePairVolume24hUSD.ts (hunk 5 above)
Comment (Diogomartf) [reply src/hooks/usePairVolume24hUSD.ts:14]: I think we could look into that, maybe create an issue?  I think it's out of scope of this PR.

**Pull request merged**